docs/source/extensions.rst

```diff
@@ -3,17 +3,233 @@
+    def initial_power(self) -> float:
+        return 16.0206
+
+Note that name of the function must match name of the argument required by the agent. If there already exists
+a function with that name, we can name the function differently and explicitly define the argument name in
+the decorator:
+
+.. code-block:: python
+
+    @parameter(parameter_name='initial_power', parameter_type=gym.spaces.Box(-np.inf, np.inf, (1,)))
+    def default_pow(self) -> float:
+        return 16.0206
+
+We define the **observation functions** analogous to parameter functions. The only differences are that we use
+the ``@observation`` decorator and that the implemented method takes additional parameters. Below is an
+example observation function that provides approximated collision probability in dense IEEE 802.11ax networks:
+
+.. code-block:: python
+
+    @observation
+    def pc(self, n_wifi, *args, **kwargs):
+        return 0.154887 * np.log(1.03102 * n_wifi)
+
+Note that the observation function can take parameters that are specified in the observation space.
+:ref:`BaseExt <base_ext>` methods will automatically pass the given observation to the function to allow
+dynamic computation of the returned value. What is important, observation methods take ``*args`` and ``**kwargs``
+as the last parameters (this is required by the internal behaviour of the ``setup_transformations`` function).
+As previously, name of the function should match name of the filled parameter, but we can specify parameter name
+and returned type in the decorator:
+
+.. code-block:: python
+
+    @observation(observation_name='pc', observation_type=gym.spaces.Box(0.0, 1.0, (1,)))
+    def collision_probability(self, n_wifi: int, *args, **kwargs) -> float:
+        return 0.154887 * np.log(1.03102 * n_wifi)
+
+Full source code of the IEEE 802.11ax extension can be found :ref:`here <https://github.com/m-wojnar/reinforced-lib/blob/main/reinforced_lib/exts/ieee_802_11_ax.py>`.
+
+Rules and limitations
+---------------------
+
+Extensions are very powerful mechanism that makes the Reinforced-lib universal and easy to use. The ``BaseExt``
+methods can handle complex and nested observation spaces, such as the
+:ref:`example ones <https://github.com/m-wojnar/reinforced-lib/blob/main/tests/extqs/base_ext_test.py>`.
+However, there are some rules and limitations that programmers and users must take into consideration:
+
+* arguments and parameters provided by the user have higher priority than default or calculated values provided
+  by the extension,
+* parameter functions cannot take any arguments (except ``self``),
+* you cannot use extension with a given agent if the agent requires a parameter that is not listed in the
+  extensions observation space or cannot be provided by an observation function - you have to add an observation
+  to the observation space, implement appropriate observation function of use the agent without any extension,
```

Comment:
...implement appropriate observation function **or** use the agent without any extension